### PR TITLE
WM-86 PDF Conversion Library

### PR DESCRIPTION
**Pull request by EdwardJB**
See [the README in this branch](https://github.com/skimit/pdf-conversion/blob/feature/wm86-pdf-conversion-library/README.md) plus notebooks for documentation.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -29,12 +29,7 @@ clear-venv:		### Removes existing Python virtual env.
 
 requirements:	### Installs all of the required project dependencies, from the existing requirements.txt file.
```
Comment (claudiorsa) [inline Makefile:30]: I am not sure if this is supposed to happen but when I run `make requirements` it uninstalls the jupyter package if I have it. I know it is not a dependency for the package but it is useful to try the demo in the jupyter notebook.

Comment (claudiorsa) [reply Makefile:30]: Maybe a:
 make demo
command to install jupyter notebook and matplotlib? Since they are required to run the demo.

File: tests/unit/pdf_conversion/helpers.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    return data
+
+
+def show_images(*paths, width=30, height=20):
+    """
+
+    Function for showing images side-by-side in Jupyter Notebooks.
+
+    """
+    import matplotlib.pyplot as plt
```
Comment (claudiorsa) [inline tests/unit/pdf_conversion/helpers.py:55]: I think matplot lib is not in the test requirements.

**Review by claudiorsa — APPROVED** [review]
Really nice repository Edward! I tested with my own pdfs and it is really capturing a very similar structure. There is one thing that I am not sure though:
- The resulting HTML files are all very similar because include a lot of <span> and <p> to create the same visual effect of the PDF. However, my concern is when we start training a ML model it will be biased towards the content inside these type of tags, because it will find a lot of them.
In other words, in my opinion, it seems that this transformation would be more suitable for a computer vision based Web-Model than a NLP based model. I might be missing the big picture here, but just wanted to share this thought which is related to the tranformations of this library.

File: tests/unit/pdf_conversion/helpers.py (hunk 2 above)
Comment (EdwardJB) [reply tests/unit/pdf_conversion/helpers.py:55]: This is true, but a) the import is within the function and b) the function is never called by the tests, but only by the demo notebook.

This is a bit confusing/ugly I realise, especially as this is inside the `tests` module, and stems from wanting tests/demos to share fixtures and functions. But there should probably be a set place for demo-only helpers. Something to consider for our forthcoming formal set of standards!

File: Makefile (hunk 1 above)
Comment (EdwardJB) [reply Makefile:30]: > Really nice repository Edward! I tested with my own pdfs and it is really capturing a very similar structure. There is one thing that I am not sure though:
> 
> * The resulting HTML files are all very similar because include a lot of `<span>` and `<p> `to create the same visual effect of the PDF. However, my concern is when we start training a ML model it will be biased towards the content inside these type of tags, because it will find a lot of them.
>   In other words, in my opinion, it seems that this transformation would be more suitable for a computer vision based Web-Model than a NLP based model. I might be missing the big picture here, but just wanted to share this thought which is related to the tranformations of this library.

This is a very interesting point. But there are a couple of additional things to consider.

* This is not the "final form" of the HTML that we be used to train WM. It will be passed to the `web-rendering-wm` pipeline to be rendered first. So we will get some additional information, explicit styling features, etc. You can see a rendered PDF in WM format in this comment: https://skim.atlassian.net/browse/COL-251?focusedCommentId=10951
* The WM architecture includes a CV aspect: in additional to the HTML content, each element is rendered to a bitmap, which is passed through a CNN. The intention is to get the best of both worlds NLP and CV.
* WM also trains Document Format Embeddings as part of each element (and document) embedding e.g. with a small "vocab" of `html`, `pdf`, etc. So the model will be aware of which document format it is currently dealing with and will have a little flexibility to encode that.
* The text of each element will be encoded with a "traditional" transformer-based language model, so even if the structural information is pretty homogenous/sparse, the model will hopefully learn to emphasis the text itself.
 
 
Anyway, the hope is that, all together, this should be enough. But we will see!

Comment (EdwardJB) [reply Makefile:30]: > I am not sure if this is supposed to happen but when I run `make requirements` it uninstalls the jupyter package if I have it. I know it is not a dependency for the package but it is useful to try the demo in the jupyter notebook.

This is a good point. Even so, the items in `Makefile` are somewhat outdated, and currently in flux, so I tend to consider the `README` the "definitive" way to install.

Complicating matters is I missed out the "Running Demos" section which mentions the `demo` extra, so I've now added that!

Comment (claudiorsa) [reply Makefile:30]: > > Really nice repository Edward! I tested with my own pdfs and it is really capturing a very similar structure. There is one thing that I am not sure though:
> > 
> > * The resulting HTML files are all very similar because include a lot of `<span>` and `<p> `to create the same visual effect of the PDF. However, my concern is when we start training a ML model it will be biased towards the content inside these type of tags, because it will find a lot of them.
> >   In other words, in my opinion, it seems that this transformation would be more suitable for a computer vision based Web-Model than a NLP based model. I might be missing the big picture here, but just wanted to share this thought which is related to the tranformations of this library.
> 
> This is a very interesting point. But there are a couple of additional things to consider.
> 
> * This is not the "final form" of the HTML that we be used to train WM. It will be passed to the `web-rendering-wm` pipeline to be rendered first. So we will get some additional information, explicit styling features, etc. You can see a rendered PDF in WM format in this comment: https://skim.atlassian.net/browse/COL-251?focusedCommentId=10951
> * The WM architecture includes a CV aspect: in additional to the HTML content, each element is rendered to a bitmap, which is passed through a CNN. The intention is to get the best of both worlds NLP and CV.
> * WM also trains Document Format Embeddings as part of each element (and document) embedding e.g. with a small "vocab" of `html`, `pdf`, etc. So the model will be aware of which document format it is currently dealing with and will have a little flexibility to encode that.
> * The text of each element will be encoded with a "traditional" transformer-based language model, so even if the structural information is pretty homogenous/sparse, the model will hopefully learn to emphasis the text itself.
> 
> Anyway, the hope is that, all together, this should be enough. But we will see!

Ok! Sounds good to me!

Yes, let's hope so!
